### PR TITLE
[API] Fix CALL computation cost capture in debug tracing

### PR DESCRIPTION
**Pull request by hyunsooda**
## Proposed changes

Computation cost capturing was initiated at #2090. This PR refactored the capturing timing so that its output aligns with the gas (gasUsed) value.

```diff
>       computation: 149866328,
>       computationCost: 113150,
>       depth: 1,
>       gas: 0,
>       gasCost: 0,
>       memory: [],
>       op: "CALL",
>       pc: 0,
>       stack: [],
>       storage: {}
>   }, {
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by blukat29** [comment]
How about in `interpreter.go`, using the same strategy as `gasCopy` management?

```go
	for atomic.LoadInt32(&in.evm.abort) == 0 {
		if in.cfg.Debug {
			// Capture pre-execution values for tracing.
			logged, pcCopy, gasCopy, ccCopy = false, pc, contract.Gas, in.evm.opcodeComputationCostSum
		}

		if in.cfg.Debug {
			ccCost = in.evm.opcodeComputationCostSum - ccCopy 
			in.cfg.Tracer.CaptureState(in.evm, pc, op, gasCopy, cost, ccCost, callContext, in.evm.depth, err)
			logged = true
		}
```

**Comment by blukat29** [comment]
@hyunsooda It's a non-trivial problem. Let's deal with it later.

File: blockchain/vm/logger.go [hunk 1]
```diff
@@ -149,7 +149,7 @@ func (l *StructLogger) CaptureStart(env *EVM, from common.Address, to common.Add
 // CaptureState logs a new structured log message and pushes it out to the environment
 //
 // CaptureState also tracks SSTORE ops to track dirty values.
-func (l *StructLogger) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, depth int, err error) {
+func (l *StructLogger) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost, opcodeComputationCost uint64, scope *ScopeContext, depth int, err error) {
```
Comment (hyeonLewis) [inline blockchain/vm/logger.go:152]: Is this intended that using `opcodeComputationCost`, not `cc` like others?

File: blockchain/vm/logger.go [hunk 2]
```diff
@@ -193,18 +193,15 @@ func (l *StructLogger) CaptureState(env *EVM, pc uint64, op OpCode, gas, cost ui
 		storage = l.changedValues[contract.Address()].Copy()
 	}
 	// create a new snapshot of the EVM.
-	var (
-		remainingComputationCost = env.Config.ComputationCostLimit - env.opcodeComputationCostSum
-		opcodeComputationCost    = env.interpreter.cfg.JumpTable[op].computationCost
-	)
+	remainingComputationCost := env.Config.ComputationCostLimit - env.opcodeComputationCostSum
```
[the remark this answers is not in the document]
Comment (hyeonLewis) [reply blockchain/vm/logger.go:196]: FYI: The suggested change is just for rough implementation for understanding. I think we could find a better way. (similar way as `gasCopy`)

**Comment by hyunsooda** [comment]
@blukat29 @hyeonLewis, Aligned with the gas cost output. Thank you.

**Comment by hyunsooda** [comment]
Description updated.

File: blockchain/vm/interpreter.go [hunk 3]
```diff
@@ -219,7 +221,7 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte) (ret []byte, err
 		}
 		if in.cfg.Debug {
 			// Capture pre-execution values for tracing.
-			logged, pcCopy, gasCopy = false, pc, contract.Gas
+			logged, pcCopy, gasCopy, ccCostCopy = false, pc, contract.Gas, in.evm.Config.ComputationCostLimit-in.evm.opcodeComputationCostSum
```
Comment (blukat29) [inline blockchain/vm/interpreter.go]:
```suggestion
			logged, pcCopy, gasCopy, ccLeftCopy = false, pc, contract.Gas, in.evm.Config.ComputationCostLimit-in.evm.opcodeComputationCostSum
```

File: blockchain/vm/interpreter.go [hunk 4]
```diff
@@ -247,6 +249,7 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte) (ret []byte, err
 		if in.evm.opcodeComputationCostSum > in.evm.Config.ComputationCostLimit {
 			return nil, ErrOpcodeComputationCostLimitReached
 		}
+		cc = operation.computationCost
```
Comment (blukat29) [inline blockchain/vm/interpreter.go]:
```suggestion
		ccOpcode = operation.computationCost
```

File: blockchain/vm/logger.go [hunk 5]
```diff
@@ -113,8 +113,8 @@ type Tracer interface {
 	CaptureEnter(typ OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int)
 	CaptureExit(output []byte, gasUsed uint64, err error)
 	// Opcode level
-	CaptureState(env *EVM, pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, depth int, err error)
-	CaptureFault(env *EVM, pc uint64, op OpCode, gas, cost uint64, scope *ScopeContext, depth int, err error)
+	CaptureState(env *EVM, pc uint64, op OpCode, gas, cost, cc, ccUnit uint64, scope *ScopeContext, depth int, err error)
+	CaptureFault(env *EVM, pc uint64, op OpCode, gas, cost, cc, ccUnit uint64, scope *ScopeContext, depth int, err error)
```
Comment (blukat29) [inline blockchain/vm/logger.go]:
```suggestion
	CaptureState(env *EVM, pc uint64, op OpCode, gas, cost, ccLeft, ccOpcode uint64, scope *ScopeContext, depth int, err error)
	CaptureFault(env *EVM, pc uint64, op OpCode, gas, cost, ccLeft, ccOpcode uint64, scope *ScopeContext, depth int, err error)
```

**Review by blukat29 — COMMENTED** [review]
Can we unify the variable names?
I know it's tricky because the term "computation cost" already has "cost" in it.

I propose:
- ccLeft = Limit - Sum = Log.Computation
- ccOpcode = cost of this opcode = Log.ComputationCost

**Comment by hyunsooda** [comment]
@blukat29, Updated.

**Review by blukat29 — APPROVED** [review]

**Review by hyeonLewis — APPROVED** [review]

**Pull request merged**